Replace color_strip_alpha with check_pango_color

silverhammermba opened the pull request:
Awesome uses `awful.util.color_strip_alpha` to fix colors before they are inserted into Pango markup. This is because Pango doesn't support hex colors with transparency and would return an error.

But colors in awesome are not always hex codes, for example:

 1. Sometimes background colors are inserted into Pango markup (e.g. for the menubar cursor #140). Background colors can be [Cairo patterns](http://awesome.naquadah.org/doc/api/modules/gears.color.html#create_pattern), which Pango does not support and which `color_strip_alpha` definitely won't fix.
 2. Pango also supports [X11 color names](http://en.wikipedia.org/wiki/X11_color_names), which is why `theme.fg_normal = "red"` works. Since `color_strip_alpha` assumes hex color codes it breaks any color name that happens to be 9 characters long. Try setting `theme.fg_normal = "chocolate"`, for example.

This PR replaces `color_strip_alpha` with a more robust `check_pango_color` function that returns `nil` if the color is not safe to use in Pango markup (thus allowing for replacement with fallback colors rather than showing error messages or `<Invalid text>`). This will cause a slight regression for any theme that relies on the alpha being stripped from hex colors (those colors will be replaced by the fallbacks).

actionless commented on lib/awful/util.lua.in:
```diff
@@ -46,14 +48,13 @@ function util.deprecate(see)
+-- @return True if the color valid, else nil.
+function util.check_pango_color(color)
+    if not color then return nil end
+    local attr, _ = Pango.parse_markup("<span background=\"" .. color .. "\"></span>", -1, 0)
```
if color is not string it will throw an exception, mb better to wrap it with pcall or check type first?

silverhammermba replied:
I thought about that, but it looks to me like most code in awesome doesn't type check and just assumes that colors are either strings or nil, so that's what I accounted for.

actionless replied:
but in your original bug report you were providing non-string value there, isn't it?

silverhammermba replied:
No, I was providing PNG pattern strings e.g. `"png:bg_normal.png"`

actionless commented on lib/awful/widget/tasklist.lua.in:
```diff
@@ -27,13 +27,13 @@ tasklist.filter = {}
+    local fg_normal = util.check_pango_color(args.fg_normal or theme.tasklist_fg_normal or theme.fg_normal) or "white"
     local bg_normal = args.bg_normal or theme.tasklist_bg_normal or theme.bg_normal or "#000000"
-    local fg_focus = args.fg_focus or theme.tasklist_fg_focus or theme.fg_focus
+    local fg_focus = util.check_pango_color(args.fg_focus or theme.tasklist_fg_focus or theme.fg_focus)
```
no any fallback values here and later

silverhammermba replied:
Those all fall back to `fg_normal` in the if-elses later. That is the current behavior as well.

actionless commented on lib/awful/prompt.lua.in:
```diff
@@ -152,7 +152,10 @@ local function prompt_text_with_cursor(args)
+    local cursor_color = util.check_pango_color(args.cursor_color) or "black"
+    local text_color = util.check_pango_color(args.text_color) or "black"
+
+    ret = _prompt .. text_start .. "<span background=\"" .. cursor_color .. "\" foreground=\"" .. text_color .. "\" underline=\"" .. underline .. "\">" .. char .. "</span>" .. text_end .. spacer
```
also imho it will be better for readability to split the lines

actionless commented:
one more thing, can it be a breaking change for people who are using colors with alpha part?

silverhammermba commented:
If you are using foreground color with alpha it will fall back to the default. Regrettable, but easy to fix. The alternative would require actually parsing the colors.

psychon commented:
I think some parts in awesome use Pango markup unnecessarily and should instead be switched to supporting cairo colors. I don't remember which ones and I don't remember if there was anything "unfixable".

For this PR:

Would it make sense to have check_pango_color() always return a string (and never `nil`)? It could accept an optional "default" argument which is used for invalid colors and which defaults to "black". IMO this would make things nicer for the callers.

I would insert a `tostring()` call into check_pango_color() to make it more robost.

Elv13 commented:
Fixable:
https://github.com/awesomeWM/awesome/blob/master/lib/awful/widget/tasklist.lua.in#L79
https://github.com/awesomeWM/awesome/blob/master/lib/awful/widget/taglist.lua.in#L101
https://github.com/awesomeWM/awesome/blob/master/lib/menubar/init.lua.in#L89

Harder:
The prompt

silverhammermba commented:
@psychon I would argue that `check_pango_color(x) or "black"` is more clear than `check_pango_color(x, "black")`, since you don't need to look inside the function to understand that it's substituting a default argument.

@Elv13 I haven't looked into Cairo's font stuff yet. That should be a separate PR, no?

psychon commented:
@silverhammermba I guess everyone defaults to black, so I would suggest `check_pango_color(x)` to do the same thing than `check_pango_color(x, "black")`. And in this case IMO the first version looks better (perhaps it should then be called something else than "check"? I'm not sure). If someone wants a red default, they can do `check_pango_color(x, "red")`.

Hm... in the patch there are some cases which do not want a default value. Perhaps have two different functions, one with and another without a default?

silverhammermba commented:
I think I have addressed all of the comments in the thread.

Definitely the biggest change is `lib/awful/widget/tasklist.lua`. As @psychon pointed out, some colors had defaults while others did not. But upon further investigation I found that even the colors without defaults did _eventually_ get defaulted - usually to `fg_normal` or `bg_normal` or black or white. So I moved all of those defaults which were scattered throughout the logic to the initial assignments.

However I did make a small change in behavior: previously, default values would be assigned slightly differently depending on whether the client was focused/minimized/urgent/etc. For example, if you set `fg_urgent` but not `bg_urgent`, `bg_urgent` would simply default to black. But if you set `fg_minimize` and not `bg_minimize`, _both_ colors would default to the `_normal` ones. To me this inconsistency was surprising and seemed unintentional, so I made it so that all client states are handled identically color-wise.

psychon commented:
Some of the commits should be squashed or reordered, I guess.

Also, it seems like pango does not support alpha in its colors which is why previously `strip_alpha` was used. This change will break configs by silently ignoring colors that previously (when their alpha part was still stripped) worked fine. Perhaps `ensure_pango_color` should try splitting the alpha part if the string "itself" didn't work?

I don't like merge commits in pull requests, sorry. (I guess this can be fixed up later)

The rest seems nice. Let's see what others say.

silverhammermba commented:
Yeah, I'll definitely clean up the commits once we're done tweaking it.

I pointed out the regression with color alpha in my original comment, but I'm not convinced that stripping the alpha is the right approach. Stripping the alpha seems like it could be _more_ confusing for users, since the resulting color will be close to what they specified, but not exactly right. Replacing it with the default placeholder (black) would make it more clear that their theme is using invalid fg colors (which they can easily fix in their theme).

blueyed commented on lib/awful/util.lua.in:
```diff
@@ -47,14 +49,16 @@ function util.deprecate(see)
-    end
-    return color
+-- @param fallback The color to return if the first is invalid. (default: black)
+-- @return color if it is valid, else fallback.
```
You could use `@tparam <type> ...` and `@treturn <type>` here.

blueyed commented on lib/awful/util.lua.in:
```diff
@@ -47,14 +49,16 @@ function util.deprecate(see)
+function util.ensure_pango_color(color, fallback)
+    color = tostring(color)
+    fallback = fallback or "black"
+    local attr, _ = Pango.parse_markup("<span background=\"" .. color .. "\"></span>", -1, 0)
```
Isn't there a lighter method to test if the color is valid?

blueyed commented:
Apart from my (small) remarks: :+1:

silverhammermba commented:
@blueyed I fixed up the code according to your comments. If there are no more fixes to make, I'll squash everything down and get rid of the merge commits.

Elv13 commented:
+1

blueyed commented:
:+1:

silverhammermba commented:
Squashed and ready to go.

blueyed commented:
Thanks!
